igcollect/ping.py

```diff
@@ -0,0 +1,137 @@
+                             '(number has to be bigger '
+                             'than 1 and smaller than 60)')
+    parser.add_argument('-p', '--prefix',
+                        default='ping',
+                        help='the path to the value in Graphite ')
+    parser.add_argument('-t', '--timeout', type=int, default=500,
+                        help='Initial target timeout in milliseconds. '
+                             'the amount of time that program '
+                             'waits for a response')
+    parser.add_argument('-d', '--delay', type=int, default=1800,
+                        choices=range(0,5000),
+                        help='time to wait between a series of pings in '
+                             'milliseconds')
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+
+    chunked_hosts = [args.hosts[i:i + 10]
+                     for i in range(0, len(args.hosts), 10)]
+
+    data = [(args.prefix, chunk, args.count, args.timeout, args.delay)
+            for chunk in chunked_hosts]
+
+    with Pool(cpu_count() * 2) as p:
+        p.starmap(check_pings, data)
+
+
+def check_pings(prefix, hosts, count, timeout, delay):
+    values = pings(hosts, count, timeout, delay)
+
+    timestamp = str(int(time.time()))
+
+    for data in values:
+        print_ping(prefix, data, timestamp)
+
+
+def pings(hosts, count, timeout, delay):
+    hosts_string = ' '.join(hosts)
+    cmd = 'fping -B1 -q -C {} -p {} -t {} {}'
+    cmd = cmd.format(count, delay, timeout, hosts_string)
+    # output:
+    # example.com : 105.94 - 104.78
+    # Showing times for each ping. A ping timeout will be shown as '-'.
+    output = subprocess.getoutput(cmd).split('\n')
+
+    values = []
+    for line in output:
+        data = {}
+        parts = line.split(':', 1)
+        data['dest'] = parts[0].strip(' ').replace('.', '_')
+        pings = parts[1].split()
+
+        total = 0.0
+        fails = 0
+        for i, ping in enumerate(pings):
+            if ping == '-':
+                pings[i] = '-1'
+                fails += 1
+            else:
+                total += float(ping)
+
+        pings = list(map(float, pings))
+
+        data['max'] = max(pings)
+        data['min'] = min(pings)
+
+        data['pings'] = pings
+        if (len(pings) <= fails):
+            data['avg'] = -1
+        else:
+            avg = total / (len(pings) - fails)
+            data['avg'] = avg
+
+        data['fails'] = fails
+
+        data['std'] = stdev(pings)
+
+        values.append(data)
+
+    return values
+
+
+def print_ping(prefix, data, timestamp):
+    template = (
+        str(prefix) + '.' + data['dest'] + '.{} {} ' + timestamp
+    )
+
+    for num, ping in enumerate(data['pings'], 1):
+        print(template.format('ping' + str(num), ping))
+
+    print(template.format('fails', data['fails']))
+
+    if len(data['pings']) == 0:
+        raise Exception('something went horrible wrong: check your fping')
```

Comment:
**First**:
```suggestion
        raise Exception('Something went horribly wrong: Check your fping!')
```

**Second**: Do not use so generic Exceptions. Do not make the user of the function in need of catching such a broad Exception. There are quite some good Exceptions already in Python which you can use for this occasion here. Make use of them.

**Third**: We made this function to be more of a utility function. Make the messages as part of such a utility function. The utility function does not know about fping or whatever was used. For this function the parameter which were given doesn't make sense.

**Fourth**: Actually for me it does makes sense. One could also just print a part of the data and just skip the other one. So for me if would make sense to just skip the print if the length is `0`.